sslcontext-kickstart-for-pem/src/main/java/nl/altindag/ssl/util/PemFormatter.java

```diff
@@ -0,0 +1,115 @@
+            String footer = certificateMatcher.group(3);
+
+            List<String> innerEncryptionHeader = extractInnerEncryptionHeaderIfPossible(body);
+            String certificateContent = extractBase64EncodedString(body, 0, String.join(EMPTY, innerEncryptionHeader));
+
+            List<String> certificateContainer = Stream.of(certificateContent.split("(?<=\\G.{64})"))
+                    .collect(Collectors.toCollection(ArrayList::new));
+            certificateContainer.add(0, header);
+            certificateContainer.addAll(1, innerEncryptionHeader);
+            certificateContainer.add(footer);
+            certificates.addAll(certificateContainer);
+        }
+
+        return certificates.isEmpty() ? value : String.join(System.lineSeparator(), certificates);
+    }
+
+    /**
+     * Extracts the inner header of a pem formatted certificate. It currently supports two encryption algorithms DES-EDE3-CBC and AES-256-CBC. See below for the examples:
+     *
+     * example
+     * <pre>
+     * {@code
+     * Proc-Type: 4,ENCRYPTED
+     * DEK-Info: DES-EDE3-CBC,9A4511D01C56B15D
+     * }
+     * </pre>
+     * or
+     * <pre>
+     * {@code
+     * Proc-Type: 4,ENCRYPTED
+     * DEK-Info: AES-256-CBC,AB8E2B5B2D989271273F6730B6F9C687
+     * }
+     * </pre>
+     */
+    private static List<String> extractInnerEncryptionHeaderIfPossible(String value) {
+        Map<String, Integer> encryptionAlgorithmsAndSaltToFieldLength = new HashMap<>();
+        encryptionAlgorithmsAndSaltToFieldLength.put("AES-256-CBC", 54);
+        encryptionAlgorithmsAndSaltToFieldLength.put("DES-EDE3-CBC", 39);
+
+        if (!value.contains(INNER_ENCRYPTED_HEADER)) {
+            return Collections.emptyList();
+        }
+
+        for (Map.Entry<String, Integer> encryptionAlgorithmAndSaltToFieldLength : encryptionAlgorithmsAndSaltToFieldLength.entrySet()) {
+            if (value.contains(encryptionAlgorithmAndSaltToFieldLength.getKey())) {
+                String encryptionAlgorithmValue = value.substring(INNER_ENCRYPTED_HEADER.length(), INNER_ENCRYPTED_HEADER.length() + encryptionAlgorithmAndSaltToFieldLength.getValue());
+                List<String> innerHeader = new ArrayList<>();
+                innerHeader.add(INNER_ENCRYPTED_HEADER);
+                innerHeader.add(encryptionAlgorithmValue);
+                innerHeader.add(EMPTY);
+                return innerHeader;
+            }
+        }
+
+        return Collections.emptyList();
+    }
+
+    private static String extractBase64EncodedString(String value, int startIndex, String innerEncryptionHeader) {
+        try {
+            String substring = value.substring(innerEncryptionHeader.length() + startIndex);
+            Base64.getDecoder().decode(substring);
+            return substring;
+        } catch (IllegalArgumentException exception) {
+            return extractBase64EncodedString(value, ++startIndex, innerEncryptionHeader);
```

Comment:
If no correct value is found this loops until the string is out indexed. Might return a weird error?

Reply:
This part was a work-in-progress when I needed to extract the inner body when an encryption header was present. I have now an alternative method, so this whole method is not needed any more, so I removed it.

Reply:
Nice, I also like the quick return in `extractInnerEncryptionHeaderIfPossible`. Java isn't so smart to initialize the hashmap after the potential quick return I think. But I could be wrong there.

Reply:
Yeah, it is not so nice to recreate the map if it just only contains the those constants and never gets updates. It is a bit of a waste to recreate the map when the method is getting called. I refactored it to a constant now.